src/NUnitEngine/nunit.engine/Runners/WorkItemTracker.cs

```diff
@@ -0,0 +1,113 @@
+        private ManualResetEvent _allItemsComplete = new ManualResetEvent(false);
+        private object _trackerLock = new object();
+        
+        public void Clear()
+        {
+            lock (_trackerLock)
+            {
+                _itemsInProcess.Clear();
+                _allItemsComplete.Reset();
+            }
+        }
+
+        public bool WaitForCompletion(int millisecondsTimeout)
+        {
+            return _allItemsComplete.WaitOne(millisecondsTimeout);
+        }
+
+        public void IssuePendingNotifications(ITestEventListener listener)
+        {
+            lock (_trackerLock)
+            {
+                int count = _itemsInProcess.Count;
+
+                // Signal completion of all pending suites, in reverse order
+                while (count > 0)
+                    listener.OnTestEvent(CreateNotification(_itemsInProcess[--count]));
+            }
+        }
+
+        private static string CreateNotification(XmlNode startElement)
+        {
+            bool isSuite = startElement.Name == "start-suite";
+
+            XmlNode notification = XmlHelper.CreateTopLevelElement(isSuite ? "test-suite" : "test-case");
+            if (isSuite)
+                notification.AddAttribute("type", startElement.GetAttribute("type"));
+            notification.AddAttribute("id", startElement.GetAttribute("id"));
+            notification.AddAttribute("name", startElement.GetAttribute("name"));
+            notification.AddAttribute("fullname", startElement.GetAttribute("fullname"));
+            notification.AddAttribute("result", "Failed");
+            notification.AddAttribute("label", "Cancelled");
+            XmlNode failure = notification.AddElement("failure");
+            XmlNode message = failure.AddElementWithCDataSection("message", "Test run cancelled by user");
+            return notification.OuterXml;
+        }
+
+        void ITestEventListener.OnTestEvent(string report)
+        {
+            XmlNode xmlNode = XmlHelper.CreateXmlNode(report);
+
+            lock (_trackerLock)
+            {
+                switch (xmlNode.Name)
+                {
+                    case "start-test":
+                    case "start-suite":
+                        _itemsInProcess.Add(xmlNode);
+                        break;
+
+                    case "test-case":
+                    case "test-suite":
+                        string id = xmlNode.GetAttribute("id");
+                        RemoveItem(id);
+
+                        if (_itemsInProcess.Count == 0)
+                            _allItemsComplete.Set();
+                        break;
+                }
+            }
+        }
+
+        private void RemoveItem(string id)
+        {
+            foreach (XmlNode item in _itemsInProcess)
+            {
+                if (item.GetAttribute("id") == id)
+                {
+                    _itemsInProcess.Remove(item);
+                    return;
```

Comment:
How about some sort of id-based hashset here? There's probably going to be single figure numbers of tests running at any one point, but still probably quite an intensive bit of code, I'd think?

Reply:
In fact, my thinking was that it's likely to not be very deep. I figured a hash or dictionary would have to be coupled with a list, as we send the notifications in reverse order. So that would end up being even more overhead.

The biggest overhead load will occur when someone has a very large number of very fast tests.

Also, I'm not very sure how niche this is. I have yet to see an NUnit framework that cancels properly. And typically, the largest number of users don't upgrade very fast.